CMakeLists.txt

```diff
@@ -1,3 +1,4 @@
+OPTION(USE_PRECISE_TEXT_PARSER "Use precise double parser for text input file" OFF)
```

Comment:
Why do we need new compilation option only for one function? Why not simply use `AtofPrecise` instead of `Atof` by default?
Having a lot of functions doing the same work is very confusing, greatly increases maintenance burden and hurts overall development process.

Reply:
1. If `AtofPrecise` is faster (or not much slower) than `Atof`, I'd like to use it by default. But it's actually much slower in my simple benchmark
    - When I see `fast_double_parser` mentioned in the commit log, I thought it's been used for text file parsing too. But it's actually not which confused me at first. I guess someone might have done the performance test and thus not using it for text parsing
    - With this performance difference, I'd choose precise version only when precision is required
2. `AtofPrecise` does not behave exactly the same with `Atof`. For some of my test models, prediction results for csv input can differ starting from the 5th non-zero digit
    - As comment in `utils/common.h` says both `StringToHelperFast` and `StringToHelper` are kept to maintain bit-for-bit legacy LightGBM behavior for precision, I guess you'd prefer to keep the old behavior by default
    - I noticed this when working on #4089 and some followups. Result verification shows this difference which makes me upset about the correctness

I've also been biten by ClickHouse's choice of non-precise float parsing by default. They choose to sacrifice 1 or 2 bits of precision to keep good float parsing performance https://github.com/ClickHouse/ClickHouse/issues/1665 (3 years ago, not sure whether still true now). For me, I choose to change the float parsing function to precise and the compile the package myself.

Reply:
OK, I see now that the main reason is that `AtofPrecise` is much slower than current solution.
But I still strongly against adding new compilation option for this because of maintenance burden and not many users will compile the library on their own to get precise file parsing.
However, I believe that new **config param** will be a good workaround for this situation. Just like recently added (see #3494 and #3578) [`deterministic` param](https://lightgbm.readthedocs.io/en/latest/Parameters.html#deterministic). Users don't have to re-compile the library every time they are switching "performance/accuracy" scenarios but do it in the runtime.
Is it possible to add new param for precise parsing? WDYT? Thanks!

Reply:
Thanks for the feedback. I'll add a config parameter and remove the compile time option to do this.

BTW, `deterministic` parameter is very useful for verifying result when change code in LightGBM.

Reply:
Now that we have the config parameter for precise float parsing, shall we remove this compile option?